Add meta registration for _foreach_norm (2nd try)

janeyx99 opened the pull request:
The first try reused TensorListMetadata, which caused illegal memory access issues when there were too many tensors in the list. We just launch multiple kernels with a simpler version of the struct (to minimize kernels launched).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119927

janeyx99 committed:
Add meta registration for _foreach_norm (2nd try)

[ghstack-poisoned]

janeyx99 committed:
Update on "Add meta registration for _foreach_norm (2nd try)"




[ghstack-poisoned]

janeyx99 commented on aten/src/ATen/native/cuda/ForeachReduceOp.cu:
```diff
@@ -200,11 +223,28 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
-              max_chunks_per_tensor);
-          C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+          const size_t num_kernels = ceil_div(ntensors, MAX_TENSORS_PER_KERNEL);
+          for (auto i = 0; i < num_kernels; i++) {
+            const size_t num_tensors_this_kernel =
+                (i < num_kernels - 1 || ntensors % MAX_TENSORS_PER_KERNEL == 0)
+                ? MAX_TENSORS_PER_KERNEL
+                : (ntensors % MAX_TENSORS_PER_KERNEL);
+
+            TensorListAddresses addr_struct;
+            for (auto j = 0; j < num_tensors_this_kernel; j++) {
+              addr_struct.addresses[j] = vec_res[i * MAX_TENSORS_PER_KERNEL + j]
+                                             .mutable_data_ptr<scalar_t>();
+            }
```
Not sure if there is a way to abstract this common code to its own function due to needing to know what `scalar_t` is

ngimel replied:
you could have a function templated on NormType that would call `AT_DISPATCH_FLOATING_TYPES_AND2`

albanD commented on aten/src/ATen/native/cuda/ForeachReduceOp.cu:
```diff
@@ -178,9 +196,14 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
   const auto options = tensors[0].options();
   auto output_per_tensor = at::zeros(
-      {ntensors * max_chunks_per_tensor},
+      {static_cast<long>(ntensors) * max_chunks_per_tensor},
```
```suggestion
      {static_cast<uint64_t>(ntensors) * max_chunks_per_tensor},
```
?

janeyx99 replied:
ah

albanD commented on test/test_foreach.py:
```diff
@@ -646,6 +646,20 @@ def test_foreach_l2_large_value_input(self, device, dtype, op):
 
+    @onlyCUDA
+    @ops(foreach_reduce_op_db, allowed_dtypes=floating_types())
+    def test_foreach_norm_big_num_tensors(self, device, dtype, op):
```
Why is this only running foreach_norm give that it is decorated with "foreach_reduce_op_db"?

janeyx99 replied:
foreach_norm is our only reduce op but yes i see your point + will remove!

janeyx99 committed:
Update on "Add meta registration for _foreach_norm (2nd try)"




[ghstack-poisoned]

albanD reviewed (approved):
SGTM!

janeyx99 committed:
Update on "Add meta registration for _foreach_norm (2nd try)"

The first try reused TensorListMetadata, which caused illegal memory access issues when there were too many tensors in the list. We just launch multiple kernels with a simpler version of the struct (to minimize kernels launched).




[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/ForeachReduceOp.cu:
```diff
@@ -223,11 +263,28 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
+
+          const size_t num_kernels = ceil_div(ntensors, MAX_TENSORS_PER_KERNEL);
+          for (auto i = 0; i < num_kernels; i++) {
+            const size_t num_tensors_this_kernel =
+                (i < num_kernels - 1 || ntensors % MAX_TENSORS_PER_KERNEL == 0)
+                ? MAX_TENSORS_PER_KERNEL
+                : (ntensors % MAX_TENSORS_PER_KERNEL);
```
```
const size_t num_tensors_this_kernel = std::min(MAX_TENSORS_PER_KERNEL, ntensors - i * MAX_TENSORS_PER_KERNEL)
```

ngimel reviewed (commented):
looks good, minor nits
